### ISSUE TITLE
Support for React Router

### ISSUE THREAD

**Issue by JohnStrick** [cypress-io/cypress#2059]
### Current behavior:
I'm prototyping using Cypress to test our app. But I ran into difficulty right away with the the location URL not being updated in the test runner. Our app uses React Router V4 to handle navigations in our single-page-app. This means we aren't actually doing requests to the server for each URL change. Rather, we are updating the URL to give the appearance that the user is on a new page, but behind the scenes we are only making requests for data to the server and not an actual request for the page in the browser's URL.

So when I run the a test like this:
    
```js
    const nextButtonSelector = 'button[type=submit]';

    cy.visit('https://accounts.outreach.io/users/sign_in')
    cy.get('#user_email').type('username');
    cy.get(nextButtonSelector).click();
    cy.get('input[type=password]').type('passord!');
    cy.get(nextButtonSelector).click();
    cy.location('pathname').should('include', '360');
```

It logs-in just fine, but it fails to find the "360" page in the URL. This is not actually a URL that we issue an HTTP request for. Rather, it's a react-router "route" that should show in the browser's URL and then we completely client-side render the page in React.

### Desired behavior:
The URL in the Cypress test runner should update just like the URL of the normal browser window when it is not being controlled by Cypress.

### Steps to reproduce:
I'd like to know if anyone else has had Cypress running via React-Router. If not, I'm happy to help set-up a small app using React Router so that there's an easy repro environment. I just would like to know first if this is a new feature or just a bug fix first before going through that effort.

### Versions
Cypress 3.0.1
Chrome 67.0.3396.99 
Mac OS X 10.13.4 (17E202)

**Comment by JohnStrick** [comment]
I don't see any error thrown. I just don't see the URL being updated. When the code executes `cy.location('pathname').should('include', '360');` it returns false because the URL is not updated. I'll see if I can get something setup in a few days to have a more localized repro.

**Comment by brian-mann** [comment]
You don't see any error thrown? Cypress will automatically retry the assertion until they pass. It's not possible for the assertion **not** to fail but also not to pass.

Can you take a screenshot of both the command log and your URL. What you're saying isn't making any sense to me.

**Comment by JohnStrick** [comment]
Sorry, I think I was confused by the terminology. The assertion does indeed fail, but I didn't see anything "thrown" in the console. Here's a screenshot.

![image](https://user-images.githubusercontent.com/36967268/42063780-29aa72a8-7ae8-11e8-962a-e19748aff38b.png)

After the log-in page redirects to our SPA, the URL never changes again. (It should automatically navigate to /360.) And even if I manually click on other links after the test finishes executing, the page content changes correctly, but the browser URL doesn't change at all. This doesn't happen when navigating to our app in a normal Chrome window.

I'm going to try and debug our app while the test is executing to see if I can find out more information. Something really strange is going on here.

**Comment by JohnStrick** [comment]
I think I found something weird that our app is doing. We think we are running in some other mode than the default mode (probably because of how Cyprus is hosting the app). I'm going to debug a little while more, but I think this is related to our product and not Cyprus.

**Comment by JohnStrick** [comment]
Yes, it's definitely our product. We have logic that looks to see if we are "windowed" and then we do special routing logic. And since Cyprus hosts our app in some kind of iFrame, we go into this special mode. I'll fix that and try again to prototype some sample Cyprus tests. Sorry for the false alarm.

**Comment by bahmutov** [comment]
You can check from you app if you are running inside Cypress test using 

if (window.Cypress) 
  ... avoid window logic

**Comment by brian-mann** [comment]
Also we have a lot of code that removes obstructive code that your app may use to figure out whether its in an iframe.

We actually fixed an issue that may be affecting this that's releasing in `3.0.2`. However, if you could - please post the code that your app uses to figure out whether its in an iframe.

Also, make sure that you haven't disabled `modifyObstructiveCode` in your `cypress.json`.

**Comment by JohnStrick** [comment]
Nice, the `window.Cypress` check will be easy to implement. I'll give that a shot.

Here's our logic: `window.self !== window.top`. That's how we were figuring out if we were in a stand-alone browser our hosted in another application/window.

**Comment by brian-mann** [comment]
Yup that is the exact thing we fixed in `3.0.2` that's about to go out. After you upgrade you shouldn't need to change any code - it will "just work". That was the one edge case we missed.

https://github.com/cypress-io/cypress/issues/2030

**Comment by JohnStrick** [comment]
That's awesome! Thanks a lot for all your help. Really appreciate it.

**Comment by brian-mann** [comment]
Released in [`3.0.2`](https://docs.cypress.io/guides/references/changelog.html#3-0-2).
